### PR TITLE
Create instance on GCP

### PR DESCRIPTION
**Pull request by gabrielbussolo**
# Summary of this proposal

## Issue(s) number(s)

*Disclaimer: Please do not create a Pull Request without creating an issue first.*

## Checklist for submitter

- [ ] My PR has a related issue/bug number.
- [ ] My PR provides tests
  - [ ] Integration tests are added/passes
  - [ ] Unit tests are added/passes
  - [ ] End-to-end tests added/passes
  - [ ] Static code check added/passes
  - [ ] Linting on documentation added/passes
  - [ ] Does not affect code coverage stats
- [ ] My PR requires updating dependencies
- [ ] My PR has the documentation changes required.

## Checklist for reviewer

- [ ] The proposal fixes a bug/issue or implements a new feature that is well described.
- [ ] The proposal has sufficient test cases that covers the changes.
  - [ ] If changes an API, it does not break backwards compatibility (-1 major version).
- [ ] If integration is required, the proposal has integration tests.
- [ ] The proposal does not break coverage stats
- [ ] The proposal has the required documentation changes.

## Backport

- [ ] Backport to the latest stable release.

### CODE REVIEW

File: cmd/calyptia/create_core_instance_gcp.go [hunk 1]
```diff
@@ -1,19 +1,115 @@
...
+					log.Fatal(err)
+				}
+
+				if operation.Status == "DONE" {
+					cmd.Println("done.")
+					break
+				}
+			}
+
+			cmd.Println("[*] Calyptia Core Instance created.")
```
Comment (niedbalski) [inline cmd/calyptia/create_core_instance_gcp.go:96]: @gabrielbussolo please add the instance ip address (either public or private).

File: cmd/calyptia/create_core_instance_gcp.go [hunk 2]
```diff
@@ -1,19 +1,115 @@
...
+
+			for {
+				operation, err := client.FollowOperations()
+
+				if err != nil || operation.Error != nil {
+					client.Rollback()
+					log.Fatal(err)
+				}
+
+				if operation.Status == "DONE" {
```
Comment (niedbalski) [inline cmd/calyptia/create_core_instance_gcp.go:90]: convert this into a const.

File: gcp/config.go [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+		projectID: projectID,
+		Resources: []resources{
+			{
+				Type: "compute.v1.instance",
+				Name: "calyptia-core-instance",
+				Properties: properties{
+					Zone:        DefaultZone,
+					MachineType: fmt.Sprintf("projects/%s/zones/%s/machineTypes/%s", projectID, DefaultZone, DefaultMachineType),
+					Metadata:    metadata{Items: []item{}},
+					Tags:        tags{Items: []string{"calyptia-core-instance"}},
```
Comment (niedbalski) [inline gcp/config.go:99]: Tags has to be extended with 2 things: 1 ) instance name 2) environment name, so you can filter out later when deleting.

File: cmd/calyptia/create_core_instance_gcp.go [hunk 4]
```diff
@@ -1,19 +1,115 @@
...
 		RunE: func(cmd *cobra.Command, args []string) error {
-			return errors.New("not implemented")
+			ctx := context.Background()
+			gcpImage, err := index.NewGCP()
+			if err != nil {
+				return err
+			}
+
+			if coreInstanceVersion != "" {
+				coreInstanceVersion, err = gcpImage.Match(ctx, coreInstanceVersion, "us")
```
Comment (niedbalski) [inline cmd/calyptia/create_core_instance_gcp.go:39]: please make the zone/region parametrised.

**Review by niedbalski — CHANGES_REQUESTED** [review]
Overall looks good, however there are a few improvements required (noted in the code).

File: aws/client.go [hunk 5]
```diff
@@ -614,6 +614,7 @@ func (c DefaultClient) CreateUserdata(in *CreateUserDataParams) (string, error)
 	return base64.StdEncoding.EncodeToString(out.Bytes()), nil
 }
 
+//CreateInstance(ctx context.Context, in *CreateInstanceParams) (CreatedInstance, error)
```
Comment (nicolasparada) [inline aws/client.go:617]: Is this a code comment?
Old code?

File: cmd/calyptia/create_core_instance_gcp.go [hunk 6]
```diff
@@ -1,19 +1,115 @@
 package main
 
 import (
-	"errors"
-
+	"context"
+	"fmt"
+	"github.com/calyptia/cli/gcp"
+	"github.com/calyptia/core-images-index/go-index"
```
Comment (nicolasparada) [inline cmd/calyptia/create_core_instance_gcp.go:7]: Missing a import sort/group here.

```bash
goimports -w -local github.com/calyptia/ .
```

File: cmd/calyptia/create_core_instance_gcp.go [hunk 7]
```diff
@@ -1,19 +1,115 @@
...
+		externalIP            string
+		sshKeyPath            string
+		sshUser               string
+	)
 	cmd := &cobra.Command{
 		Use:     "gcp",
 		Aliases: []string{"google", "gce"},
-		Short:   "Setup a new core instance on Google Compute Engine (TODO)",
+		Short:   "Setup a new core instance on Google Compute Engine",
+		Long:    "Setup a new core instance on Google Compute Engine, you need to be authenticated with gcloud cli \n",
```
Comment (nicolasparada) [inline cmd/calyptia/create_core_instance_gcp.go:30]:
```suggestion
		Long:    "Setup a new core instance on Google Compute Engine, you need to be authenticated with gcloud cli",
```

File: cmd/calyptia/create_core_instance_gcp.go [hunk 8]
```diff
@@ -1,19 +1,115 @@
...
 		RunE: func(cmd *cobra.Command, args []string) error {
-			return errors.New("not implemented")
+			ctx := context.Background()
```
Comment (nicolasparada) [inline cmd/calyptia/create_core_instance_gcp.go:32]: Recently, I noticed you can get the context from `cmd`.
```suggestion
			ctx := cmd.Context()
```

File: cmd/calyptia/create_core_instance_gcp.go [hunk 9]
```diff
@@ -1,19 +1,115 @@
...
+			}
+
+			if coreInstanceVersion != "" {
+				coreInstanceVersion, err = gcpImage.Match(ctx, coreInstanceVersion, "us")
+				if err != nil {
+					return err
+				}
+			}
+
+			if err != nil {
```
Comment (nicolasparada) [inline cmd/calyptia/create_core_instance_gcp.go:45]: You left this here :p

File: gcp/client.go [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+func (c *DefaultClient) SetConfig(newConfig Config) {
+	c.config = newConfig
+}
+
+func (c *DefaultClient) Deploy() error {
+	configBytes, err := yaml.Marshal(c.config)
+	if err != nil {
+		return err
+	}
+	fmt.Println(string(configBytes))
```
Comment (nicolasparada) [inline gcp/client.go:38]:
```suggestion
```

File: gcp/client.go [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+func (c *DefaultClient) FollowOperations() (*deploymentmanager.Operation, error) {
+	getOperation := c.service.Operations.Get(c.projectName, c.deploymentName)
+	return getOperation.Do()
+}
+
+func (c *DefaultClient) Rollback() {
+	if err := c.Delete(); err != nil {
+		log.Fatal(err)
```
Comment (nicolasparada) [inline gcp/client.go:76]:
```suggestion
```
That is bad.

File: gcp/client.go [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+		return err
+	}
+	c.deploymentName = insertDeployment.Name
+	return nil
+}
+
+func New(ctx context.Context, projectName string, environment string) (*DefaultClient, error) {
+	service, err := deploymentmanager.NewService(ctx)
+	if err != nil {
+		log.Fatal(err)
```
Comment (nicolasparada) [inline gcp/client.go:61]:
```suggestion
		return nil, err
```

File: gcp/config.go [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+				Value: "$(ref.calyptia-core-instance.networkInterfaces[0].accessConfigs[0].natIP)",
+			},
+		},
+	}
+	return c
+}
+
+func (c *Config) SetZone(zone string) *Config {
+	if zone == "" {
+		return nil
```
Comment (nicolasparada) [inline gcp/config.go:136]: That is an awful side-effect.
You could try returning an error instead.
If you prefer the fluent API design here, try holding an error in the config and return it at the end. Something like:
```go
cfg := NewConfig()
err := cfg.SetA(a).
  SetB(b).
  SetC(c).
  SetD(d).
  Build()

```

File: gcp/config.go [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+	}
+	c.Resources[0].Properties.Zone = zone
```
Comment (nicolasparada) [inline gcp/config.go:138]: This could panic. Missing a nigl check.

File: gcp/config.go [hunk 15]
```diff
@@ -0,0 +1,240 @@
...
+	return c
+}
+
+func (c *Config) SetMachineType(machineType string) *Config {
+	if machineType == "" {
+		return nil
+	}
+	c.Resources[0].Properties.MachineType = fmt.Sprintf("projects/%s/zones/%s/machineTypes/%s", c.projectID, c.Resources[0].Properties.Zone, machineType)
```
Comment (nicolasparada) [inline gcp/config.go:146]: Missing index accesor check on `Resources[0]`.

File: gcp/config.go [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+	if zone == "" {
+		return nil
+	}
+	c.Resources[0].Properties.Zone = zone
+	return c
+}
+
+func (c *Config) SetMachineType(machineType string) *Config {
+	if machineType == "" {
+		return nil
```
[the remark this answers is not in the document]
Comment (nicolasparada) [reply gcp/config.go:144]: Actually, its everywhere. Please fix it in all places.

File: gcp/config.go [hunk 17]
```diff
@@ -0,0 +1,240 @@
...
+		return
+	}
+	c.Resources[0].Properties.NetworkInterfaces[0].AccessConfigs[0].NatIP = ip
+
+}
+
+func (c *Config) loadKey(path string) string {
+	file, err := os.ReadFile(path)
+	if err != nil {
+		log.Fatal("Could not load key: ", err)
```
Comment (nicolasparada) [inline gcp/config.go:237]: Avoid panics.
Return an error or even return an empty string, but don't panic.

**Review by nicolasparada — CHANGES_REQUESTED** [review]
Please add context argument to the gcp client.
Fix some import sorting/group that can be cone with `goimports.`
Avoid panics (log.Fatal/insecure index accessing/nil returns).

File: gcp/config.go (hunk 14 above)
Comment (gabrielbussolo) [reply gcp/config.go:138]: im not checking this index bcz is the one that we create by default, so dont will have less than [0], you think we still need to check?

File: gcp/config.go (hunk 13 above)
Comment (gabrielbussolo) [reply gcp/config.go]: i modified now if u can review again

File: gcp/config.go [hunk 18]
```diff
@@ -0,0 +1,240 @@
...
+							Boot:       true,
+							AutoDelete: true,
+							InitializeParams: initializeParams{
+								SourceImage: BaseCalyptiaImage,
+							},
+						},
+					},
+					NetworkInterfaces: []networkInterface{
+						{
+							Network: fmt.Sprintf("https://www.googleapis.com/compute/v1/projects/%s/global/networks/default", projectID),
```
[the remark this answers is not in the document]
Comment (niedbalski) [reply gcp/config.go:112]: This isn't resolved @gabrielbussolo

**Review by niedbalski — CHANGES_REQUESTED** [review]
A couple of observations.

File: gcp/config.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (gabrielbussolo) [reply gcp/config.go:112]: https://github.com/calyptia/cli/pull/114/files#diff-70820affddd170aca86c0cf9c99cf26da1afc7d9f98dfd34b26fea40ae482dc2R71 
https://github.com/calyptia/cli/pull/114/files#diff-d5ca49b3a98b0d7e5d9953cf4d59862473a0fcc31f7a29961fcd476000d0c90dR234

File: gcp/config.go (hunk 14 above)
Comment (nicolasparada) [reply gcp/config.go:137]: Resources is an exported field so it can be set to anything. Can perfectly be nil or an empty slice.

**Review by niedbalski — APPROVED** [review]
LGTM

**Review by nicolasparada — APPROVED** [review]
I see some small details but should be good to merge.
Thanks for the hard work @gabrielbussolo :)

**Pull request merged**